src/track/beats.h

```diff
@@ -52,123 +82,236 @@ class Beats : public QObject {
+    /// given serialization with the version that produced it.
+    virtual QString getVersion() const;
+    /// Return a string that represent the preferences used to generate
+    /// the beats object.
+    virtual QString getSubVersion() const;
+    virtual void setSubVersion(QString subVersion);
+    bool isValid() const;
+    /// Calculates the BPM between two beat positions.
+    double calculateBpm(const track::io::Beat& startBeat,
+            const track::io::Beat& stopBeat) const;
+
+    /// Initializes the BeatGrid to have a BPM of dBpm and the first beat offset
+    /// of dFirstBeatSample. Does not generate an updated() signal, since it is
+    /// meant for initialization.
+    void setGrid(double dBpm, double sample = 0) {
+        setGridNew(dBpm, sample / 2.0);
+    }
+    void setGridNew(double dBpm, FrameNum dFirstBeatFrame = 0);
+
+    // TODO: We may want to implement these with common code that returns
+    //       the triple of closest, next, and prev.
+
+    /// Starting from frame, return the frame number of the next beat
+    /// in the track, or -1 if none exists. If frame refers to the location
+    /// of a beat, frame is returned.
+    virtual double findNextBeat(double sample) const {
+        return findNextBeatNew(sample / 2.0) * 2.0;
+    }
+    virtual FrameNum findNextBeatNew(FrameNum frame) const;
+
+    /// Starting from frame frame, return the frame number of the previous
+    /// beat in the track, or -1 if none exists. If frame refers to the
+    /// location of beat, frame is returned.
+    virtual double findPrevBeat(FrameNum frame) const {
+        return findPrevBeatNew(frame / 2.0) * 2.0;
+    }
+    virtual FrameNum findPrevBeatNew(FrameNum frame) const;
+
+    /// Starting from frame, fill the frame numbers of the previous beat
+    /// and next beat.  Either can be -1 if none exists.  If frame refers
+    /// to the location of the beat, the first value is frame, and the second
+    /// value is the next beat position.  Non- -1 values are guaranteed to be
+    /// even.  Returns false if *at least one* sample is -1.  (Can return false
+    /// with one beat successfully filled)
+    virtual bool findPrevNextBeats(double sample,
+            double* pPrevSample,
+            double* pNextSample) const {
+        FrameNum prev, next;
+        bool result;
+
+        result = findPrevNextBeatsNew(sample / 2.0, &prev, &next);
+        *pPrevSample = prev * 2.0;
+        *pNextSample = next * 2.0;
+        return result;
+    }
+    virtual bool findPrevNextBeatsNew(FrameNum frame,
+            FrameNum* pPrevBeatFrame,
+            FrameNum* pNextBeatFrame) const;
+
+    /// Starting from frame, return the frame number of the closest beat
+    /// in the track, or -1 if none exists.  Non- -1 values are guaranteed to be
+    /// even.
+    virtual double findClosestBeat(double sample) const {
+        return findClosestBeatNew(sample / 2.0) * 2.0;
```

Comment:
Here you implicitly make the assumption of frames with 2 channel. Please transform all those expressions to use the corresponding constants instead of hard-coding literal numbers.

Reply:
This is temporary code until I finalize the PR.
I will add comments in the temporary adaptors to make it clear.

Reply:
Done.